pkg/forwarder/transaction.go

```diff
@@ -111,24 +111,58 @@ type HTTPTransaction struct {
 	ErrorCount int
 
 	createdAt time.Time
+	// retryable indicates whether this transaction can be retried
+	retryable bool
+
+	// attemptHandler will be called with a transaction before the attempting to send the request
+	attemptHandler func(transaction *HTTPTransaction)
+	// completionHandler will be called with a transaction after it has been successfully sent
+	completionHandler func(transaction *HTTPTransaction, statusCode int, body []byte, err error)
 }
 
 // Transaction represents the task to process for a Worker.
 type Transaction interface {
-	Process(ctx context.Context, client *http.Client) error
+	Process(ctx context.Context, client *http.Client) (int, []byte, error)
 	GetCreatedAt() time.Time
 	GetTarget() string
+	IsRetryable() bool
+	OnAttempt()
+	OnComplete(statusCode int, body []byte, err error)
```

Comment:
It seems to me that this logic could be encapsulated in the `Process` method, no ? We could have different class implementing the `Transation` interface: a `HTTPTransaction`. `HTTPTransactionNoRetry` ...

This way the `Worker` class don't have to handle the extra complexity, it just run and retry transaction. And each transaction internally determine if they should be retried, call some function when `Process` is called and is finished ...

I feel that we should leverage the `Transaction` interface rather than increasing it's complexity.

WDYT ?

Reply:
My [initial commit](https://github.com/DataDog/datadog-agent/pull/5169/commits/b29b44e479280bf0a37ea291883a7ccf6dea23ae) had a version of this where everything was encapsulated inside of the transaction but I struggled to find a nice way to add the ability to handle errors with non-retryable transactions in a single place.

I can try something different though to encapsulate all of the logic in the txn.